src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -56,7 +196,35 @@ private function exit_element( $element ) {
 		parent::release_bookmark( "{$this->depth}_{$element}" );
 	}
 
-	private function opened_element() {
+	/**
+	 * @see https://html.spec.whatwg.org/#close-a-p-element
+	 * @return void
+	 */
+	private function close_p_element() {
+		$this->generate_implied_end_tags( 'P' );
+
+
+	}
+
+	/**
+	 * @TODO: Implement this
+	 *
+	 * @see https://html.spec.whatwg.org/#generate-implied-end-tags
+	 *
+	 * @param string|null $except_for_this_element Perform as if this element doesn't exist in the stack of open elements.
+	 * @return void
+	 */
+	private function generate_implied_end_tags( $except_for_this_element = null ) {
+		
+	}
+
+	/**
+	 * The current node is the bottommost node in this stack of open elements.
+	 *
+	 * @see https://html.spec.whatwg.org/#current-node
+	 * @return false|mixed|string
+	 */
+	private function current_node() {
```

Comment:
This returns a string which won't be enough in many instances.

Reply:
so fun fact I realized today: many of these parse errors are undefined. I couldn't find specific references to them in the error table. I think they get overlooked but I plan on testing them out. in the case that it's benign I don't think we need to do anything, but otherwise I think the most pragmatic option is to have several of these helper functions `throw` and catch this in the top-level methods.

Reply:
I'm still not following you. what is there in identity beyond the tag name?

I know that for reconstructing the formatting elements we have to examine also the attributes, but we can do that via the bookmark. do we need anything beyond tag name for the stack construction and closing?

Reply:
As the parser consumes tag openers, it creates elements and pushes them to the list of active formatting elements and the stack of open elements. By identity I mean  having the same object present on both lists, even if there are other ones with the same name and attributes.